perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/AuthzResolverBlImpl.java

```diff
@@ -277,82 +303,111 @@ public static boolean isAuthorizedForAttribute(PerunSession sess, ActionType act
 				for (Vo voElement : userVos) {
 					if (isAuthorized(sess, Role.GROUPADMIN, voElement)) return true;
 				}
 			}
-//			if (roles.contains(Role.FACILITYADMIN)) ; //Not allowed
+//			if (roles.containsKey(Role.FACILITYADMIN)) ; //Not allowed
 		} else if (member != null) {
 
-			if (roles.contains(Role.VOADMIN)) {
+			if (roles.containsKey(Role.VOADMIN)) {
 				if (isAuthorized(sess, Role.VOADMIN, member)) return true;
 			}
-			if (roles.contains(Role.VOOBSERVER)) {
+			if (roles.containsKey(Role.VOOBSERVER)) {
 				if (isAuthorized(sess, Role.VOOBSERVER, member)) return true;
 			}
-			if (roles.contains(Role.SELF)) {
-				if (isAuthorized(sess, Role.SELF, member)) return true;
+			if (roles.containsKey(Role.SELF)) {
+				if (roles.get(Role.SELF).contains(ActionType.READ_PUBLIC) || roles.get(Role.SELF).contains(ActionType.WRITE_PUBLIC)) return true;
+				if (roles.get(Role.SELF).contains(ActionType.READ) || roles.get(Role.SELF).contains(ActionType.WRITE)) {
+					if (isAuthorized(sess, Role.SELF, member)) return true;
+				}
+				if (roles.get(Role.SELF).contains(ActionType.READ_VO) || roles.get(Role.SELF).contains(ActionType.WRITE_VO)) {
+					Vo voFromUser = getPerunBl().getMembersManagerBl().getMemberVo(sess, member);
+					if (getPerunBl().getUsersManagerBl().getVosWhereUserIsMember(sess, sess.getPerunPrincipal().getUser()).contains(voFromUser)) return true;
+				}
 			}
-			if (roles.contains(Role.GROUPADMIN)) {
+			if (roles.containsKey(Role.GROUPADMIN)) {
 				//if principal is groupManager in vo where the member has membership
 				Vo v = getPerunBl().getMembersManagerBl().getMemberVo(sess, member);
 				if (isAuthorized(sess, Role.GROUPADMIN, v)) return true;
 			}
-//			if (roles.contains(Role.FACILITYADMIN)) ; //Not allowed
+//			if (roles.containsKey(Role.FACILITYADMIN)) ; //Not allowed
 		} else if (vo != null) {
-			if (roles.contains(Role.VOADMIN)) {
+			if (roles.containsKey(Role.VOADMIN)) {
 				if (isAuthorized(sess, Role.VOADMIN, vo)) return true;
 			}
-			if (roles.contains(Role.VOOBSERVER)) {
+			if (roles.containsKey(Role.VOOBSERVER)) {
 				if (isAuthorized(sess, Role.VOOBSERVER, vo)) return true;
 			}
-			if (roles.contains(Role.GROUPADMIN)) {
+			if (roles.containsKey(Role.GROUPADMIN)) {
 				//if Principal is GroupManager in the vo
 				if (isAuthorized(sess, Role.GROUPADMIN, vo)) return true;
 			}
-			if (roles.contains(Role.FACILITYADMIN)) {
+			if (roles.containsKey(Role.FACILITYADMIN)) {
 				// is facility manager of any vo resource
 				List<Resource> resourceList = perunBl.getResourcesManagerBl().getResources(sess, vo);
 				for (Resource res : resourceList) {
 					if (isAuthorized(sess, Role.FACILITYADMIN, res)) return true;
 				}
 			}
-			if (roles.contains(Role.SELF)) {
-				if (actionType.equals(ActionType.READ)) {
+			if (roles.containsKey(Role.SELF)) {
+				if (actionType.equals(ActionType.READ) || actionType.equals(ActionType.READ_VO) || actionType.equals(ActionType.READ_PUBLIC)) {
```

Comment:
I think that READ_VO actionType allows reading information only to users who are part of this Vo. You should get the Vos from principal and check, if one of them is matching this one.

But I am not 100% sure here. You should probably ask @licehammer about this one.